book/src/pil/types.md

```diff
@@ -0,0 +1,85 @@
+# Types
+
+The powdr-pil language has the following types:
+
+- `bool`
+- `int` (integer)
+- `fe` (field element)
+- `string`
+- tuple
+- array
+- function type
+- expression
+- identity
+
+The typing is a mix of compile-time and run-time typing. Columns are expected
+to have type `int -> fe` (i.e. a function mapping integers (row indices) to field elements).
+This is not checked, though, but if a column is evaluated as a function call
+(to retrieve a specific value in a specific row of the column), the signature `int -> fe`
+is assumed.
+
+## List of Types
+
+### Bool
+
+Type name: `bool`
+
+Booleans are the results of comparisons. They allow the following operators:
+
+- `&&`: logical conjunction
+- `||`: logical disjunction
+- `!:` logical negation:
+
+### Integer
+
+Type name: `int`
+
+Integers in powdr-pil have unlimited size. Literal numbers are always integers.
+Array index requires an integer and row indices (for example the input to a fixed
+column defined through a function) are also integers.
+
+Integers allow the following operators, whose result is always an integer:
+
+- `+`: addition
+- `-`: subtraction
+- `*`: multiplication
+- `/`: integer division rounding towards zero
+- `**`: exponentiation, the exponent needs to be non-negative and fit 32 bits
+- `%`: remainder after division TODO specify how this behaves for negative numbers
+- `&`: bit-wise conjunction
+- `|`: bit-wise disjunction
+- `<<`: bit-wise shift left, the shift amount needs to be non-negative and fit 32 bits
+- `>>`: bit-wise shift right, the shift amount needs to be non-negative and fit 32 bits
+
+In addition, the following comparison operators are allowed, the result is a boolean:
+- `<`: less-than comparison
+- `<=`: less-or-equal comparison
+- `==`: equal comparison
+- `!=`: non-equal comparison
+- `>=`: greater-or-equal comparison
+- `>`: greater-than comparison
+
+
+The exponentiation operator on field elements requires a non-negative integer as exponent:
```

Comment:
Shouldn't this go in the next subsection?

Reply:
I mean it's relevant to both integers and field elements.

Reply:
ok